src/us/kbase/workspace/test/database/provenance/CreditMetadataTest.java

```diff
@@ -0,0 +1,883 @@
+						.withResourceType(rt)
+						.build();
+				assertCreditMetadataFields(creditMetadata, entry.getValue(), rt,
+						CONTRIBUTOR_LIST, TITLE_LIST, VERSION_MAP);
+			}
+		}
+	}
+
+	@Test
+	public void buildMinimalResourceTypeStrings() throws Exception {
+		for (Map.Entry<String, ResourceType> rtes : VALID_RESOURCE_TYPE_MAP.entrySet()) {
+			for (Map.Entry<String, String> entry : VALID_PID_MAP.entrySet()) {
+				final CreditMetadata creditMetadata = CreditMetadata
+						.getBuilder(entry.getKey(), CONTRIBUTOR_LIST,
+								TITLE_LIST)
+						.withVersion(VERSION_STRING)
+						.withResourceTypeString(rtes.getKey())
+						.build();
+				assertCreditMetadataFields(creditMetadata, entry.getValue(),
+						rtes.getValue(), CONTRIBUTOR_LIST, TITLE_LIST,
+						VERSION_MAP);
+			}
+		}
+	}
+
+	// Note: since there is currently only one ResourceType, not supplying a value
+	// for the field currently results in the default (only) value being used.
+	// This behaviour will change when more resource types are added, and these
+	// tests
+	// will need to be updated.
+	@Test
+	public void buildMinimalDefaultResourceType() throws Exception {
+		final CreditMetadata creditMetadata = CreditMetadata
+				.getBuilder(IDENTIFIER_STRING, CONTRIBUTOR_LIST, TITLE_LIST)
+				.withResourceTypeString(null)
+				.withVersion(VERSION_STRING)
+				.build();
+
+		assertCreditMetadataFields(creditMetadata, IDENTIFIER_STRING, ResourceType.DATASET,
+				CONTRIBUTOR_LIST, TITLE_LIST, VERSION_MAP);
+
+		final CreditMetadata creditMetadata2 = CreditMetadata
+				.getBuilder(IDENTIFIER_STRING, CONTRIBUTOR_LIST, TITLE_LIST)
+				.withResourceType(null)
+				.withVersion(VERSION_STRING)
+				.build();
+
+		assertCreditMetadataFields(creditMetadata2, IDENTIFIER_STRING, ResourceType.DATASET,
+				CONTRIBUTOR_LIST, TITLE_LIST, VERSION_MAP);
+
+	}
+
+	@Test
+	public void buildCleanAndTrimComments() throws Exception {
+		for (final ResourceType rt : ResourceType.values()) {
+			final CreditMetadata creditMetadata = CreditMetadata
+					.getBuilder(IDENTIFIER_STRING, CONTRIBUTOR_LIST, TITLE_LIST)
+					.withResourceType(rt)
+					.withComments(COMMENTS_LIST_UNTRIMMED)
+					.withVersion(VERSION_STRING_UNTRIMMED)
+					.build();
+
+			assertCreditMetadataFields(creditMetadata, IDENTIFIER_STRING, rt,
+					CONTRIBUTOR_LIST, TITLE_LIST, VERSION_MAP,
+					COMMENTS_LIST, ELD, ELF, ELRI);
+		}
+	}
+
+	@Test
+	public void buildCleanURLTypeLicense() throws Exception {
+		for (final ResourceType rt : ResourceType.values()) {
+			for (final String licenseURL : VALID_URL_LIST) {
+				final CreditMetadata creditMetadata2 = CreditMetadata
+						.getBuilder(IDENTIFIER_STRING, CONTRIBUTOR_LIST, TITLE_LIST)
+						.withResourceType(rt)
+						.withLicense(licenseURL)
+						.withDates(DATE_LIST)
+						.build();
+				assertCreditMetadataFields(creditMetadata2, IDENTIFIER_STRING, rt,
+						CONTRIBUTOR_LIST, TITLE_LIST,
+						ImmutableMap.of(LICENSE, licenseURL),
```

Comment:
The build step might mutate the url by normalizing it - currently that's untested AFAICT

Reply:
see the test `buildCleanURLTypeLicense`

Reply:
The input to the CM and the input to the CM contents checker function are both `licenseURL`, which means the URL wasn't normalized / mutated, right? I'm not understanding how the normalization step is tested

Reply:
To further clarify, if the `.normalize()` step in converting the URLs to strings was removed from the CM class, all the tests would pass AFAICT. If the normalize is important there should be at least one test that would fail

Reply:
To be even more pedantic and annoying https://docs.oracle.com/javase/8/docs/api/java/net/URI.html#normalize--

Reply:
😖

Reply:
:+1: